### DISCUSSION TITLE
Stuck at "Sending Command..."

### DISCUSSION THREAD

**Discussion by vincent-lu** [ytausch/serverless-discord-bot#71] [Q&A] [answered]
Hi thanks for this repo I'm trying to get it working.

After I've run `sam deploy --guided`, I can see the new command `/hello` shows up in my Discord with the correct description. When I hit enter it just stucks at "Sending Command...".

I can see the request actually gets sent to Lambda and the CloudWatch log shows:
```
Running command: hello (1049129300865065071, guild 740092388915019796)
```

Do you know why is it just stuck at "Sending Command..."?

Thanks!

**Comment by ytausch** [comment]
Hi there,

I'm sorry that you have trouble setting up serverless-discord bot. Unfortunately, I cannot reproduce this issue after following the README steps. Your CloudWatch logs look alright.

Did you set your guild id in the commands directory? Otherwise, I suggest you try to deploy the stack again. Maybe you can add more logging using `console.log` in the `run` method, although this shouldn't be necessary under normal circumstances.

**Comment by vincent-lu** [comment]
Hey @ytausch thanks for getting back to me.

Yeah maybe there is something else wrong with my setup, (perhaps my Discord App setup?). I will need to dig deeper and figure out why it is happening.

Will get back to you mate.

**Comment by vincent-lu** [comment]
Hi @ytausch I've tried this again with the `console.log` you suggested. This is my `/src/commands_layer/nodejs/commands/helloCommand.js`

```js
const {SlashCommand} = require('slash-create');

module.exports = class HelloCommand extends SlashCommand {
    constructor(creator) {
        super(creator, {
            name: 'hello',
            description: 'Says "Hello there" to you',
            guildIDs: ['7400923889########']
        });
        this.filePath = __filename;
    }

    async run(ctx) {
        console.log(ctx);
        return 'Hello there';
    }
};
```

When I try to run /hello in my Discord (i.e. guide id 7400923889########) I still get stuck at "Sending Command".

The log does show the `CommandContext`, it looks like this: (with tokens & ids removed)
```
2022-12-08T09:32:31.100Z    0c998d9e-032f-49a3-9fdd-9a2520ab98b6    INFO    CommandContext {
  invokedAt: 1670491951077,
  users: Collection(0) [Map] {},
  members: Collection(0) [Map] {},
  roles: Collection(0) [Map] {},
  channels: Collection(0) [Map] {},
  messages: Collection(0) [Map] {},
  attachments: Collection(0) [Map] {},
  creator: <ref *1> SlashCreator {
    _events: Events <[Object: null prototype] {}> {
      debug: [EE],
      warn: [EE],
      error: [EE],
      rawREST: [EE]
    },
    _eventsCount: 4,
    api: SlashCreatorAPI { _creator: [Circular *1] },
    commands: Collection(1) [Map] {
      '1:7400923889########:hello' => [HelloCommand]
    },
    _componentCallbacks: Map(0) {},
    _modalCallbacks: Map(0) {},
    client: undefined,
    options: {
      agent: null,
      allowedMentions: [Object],
      defaultImageFormat: 'jpg',
      defaultImageSize: 128,
      unknownCommandResponse: true,
      handleCommandsManually: false,
      disableTimeouts: false,
      componentTimeouts: false,
      latencyThreshold: 30000,
      ratelimiterOffset: 0,
      requestTimeout: 15000,
      maxSignatureTimestamp: 5000,
      endpointPath: '/interactions',
      serverPort: 8030,
      serverHost: 'localhost',
      applicationID: '########',
      publicKey: '####'
    },
    allowedMentions: { parse: [Array] },
    requestHandler: RequestHandler {
      baseURL: '/api/v10',
      userAgent: 'DiscordBot (https://github.com/Snazzah/slash-create, 5.10.0)',
      ratelimits: {},
      globalBlock: false,
      readyQueue: [],
      _creator: [Circular *1],
      requestTimeout: 15000,
      agent: null,
      latencyRef: [Object]
    },
    server: AWSLambdaServer {
      alreadyListening: true,
      isWebserver: true,
      _handler: [Function: bound _onRequest] AsyncFunction
    }
  },
  interactionToken: '####',
  interactionID: '####',
  channelID: '####',
  guildID: '7400923889########',
  locale: 'en-GB',
  guildLocale: 'en-US',
  member: Member {
    _creator: <ref *1> SlashCreator {
      _events: [Events <Complex prototype>],
      _eventsCount: 4,
      api: [SlashCreatorAPI],
      commands: [Collection [Map]],
      _componentCallbacks: Map(0) {},
      _modalCallbacks: Map(0) {},
      client: undefined,
      options: [Object],
      allowedMentions: [Object],
      requestHandler: [RequestHandler],
      server: [AWSLambdaServer]
    },
    joinedAt: 1666089102636,
    roles: [ '77956704499505####' ],
    pending: false,
    guildID: '7400923889########',
    id: '####',
    user: User {
      _creator: [SlashCreator],
      id: '####',
      username: 'MYUSERNAME',
      discriminator: '####',
      avatar: '',
      _flags: 0,
      bot: false
    },
    mute: false,
    deaf: false,
    _permissions: '####'
  },
  user: User {
    _creator: <ref *1> SlashCreator {
      _events: [Events <Complex prototype>],
      _eventsCount: 4,
      api: [SlashCreatorAPI],
      commands: [Collection [Map]],
      _componentCallbacks: Map(0) {},
      _modalCallbacks: Map(0) {},
      client: undefined,
      options: [Object],
      allowedMentions: [Object],
      requestHandler: [RequestHandler],
      server: [AWSLambdaServer]
    },
    id: '####',
    username: 'MYUSERNAME',
    discriminator: '6201',
    avatar: '####',
    _flags: 0,
    bot: false
  },
  appPermissions: Permissions { bitfield: ####n },
  initiallyResponded: false,
  deferred: false,
  _respond: [AsyncFunction (anonymous)],
  data: {
    app_permissions: '####',
    application_id: '########',
    channel_id: '####',
    data: {
      guild_id: '7400923889########',
      id: '####',
      name: 'hello',
      type: 1
    },
    entitlement_sku_ids: [],
    guild_id: '7400923889########',
    guild_locale: 'en-US',
    id: '####',
    locale: 'en-GB',
    member: {
      avatar: null,
      communication_disabled_until: null,
      deaf: false,
      flags: 0,
      is_pending: false,
      joined_at: '2022-10-18T10:31:42.636000+00:00',
      mute: false,
      nick: null,
      pending: false,
      permissions: '',
      premium_since: null,
      roles: [Array],
      user: [Object]
    },
    token: '',
    type: 2,
    version: 1
  },
  webserverMode: true,
  commandType: 1,
  commandName: 'hello',
  commandID: '',
  options: {},
  subcommands: [],
  _timeout: Timeout {
    _idleTimeout: 2000,
    _idlePrev: [TimersList],
    _idleNext: [TimersList],
    _idleStart: 1039,
    _onTimeout: [Function (anonymous)],
    _timerArgs: undefined,
    _repeat: null,
    _destroyed: false,
    [Symbol(refed)]: true,
    [Symbol(kHasPrimitive)]: false,
    [Symbol(asyncId)]: 4,
    [Symbol(triggerId)]: 0
  }
}
```

So it does seem the DiscordHandlerFunction on Lambda received the interaction from Discord, but something went wrong after that.

Also I didn't see you mention how the Discord Application and the Bot should be set up, what `SCOPE` should the Bot be assigned with? In my case the bot has all `TEXT PERMISSIONS` under `bot` and `applications.commands`. Is my Application/Bot misconfiged?

Again thanks for the help!

**Reply by ytausch** [reply]
Hi,

your issue seems very weird to me. Regarding bot authorization, the `application.commands` scope is actually sufficient. Maybe you can try to remove your bot from the server and authorize it again? To do that, use the "OAuth2 URL Generator" in the Discord Developer Portal, and only select `application.commands`.

Note that there won't be a bot account showing up in your server. This is intended.

After that, delete the CloudFormation stack and deploy it again so that the commands are synchronized again.

**Reply by vincent-lu** [reply]
Np thanks again I will give it another go, wish me luck

**Comment by vincent-lu** [comment, marked as answer]
Hi @ytausch thanks for your help. After completely doing everything from scratch it finally worked. (deleted Discord App/Bot and recloned this repo).

Now I have one more question for you; what's the best way to develop a complex command (e.g. make axios calls to external servers, etc.) What's the best way to "run" the command locally?

**Reply by ytausch** [reply]
Hi,

if you want to run your commands locally, I recommend taking a look at the AWS Toolkit Extension for your IDE. Unfortunately, you might run into issues when trying to add the Lambda layers to your local Lambda instances. This is also why I have no experience running serverless-discord-bot locally. Sorry!

If you want to run large, complex commands, you generally just insert the relevant code into the run() method of your command.
Just keep in mind that there is maximum first response time enforced by Discord. For more information, take a look at our dependency https://github.com/Snazzah/slash-create, especially the additional notes here: https://slash-create.js.org/#/docs/main/latest/examples/lambda

**Comment by vincent-lu** [comment]
Hey thanks @ytausch I managed to do mostly what I set out to do. I basically just prototyped my code in browser's dev tool which is a bit unglamorous but worked for me.

Again thanks for your repo and all the help!
